### ISSUE TITLE
Neural Turing Machines PR accepted and here is what you can learn from it

### ISSUE THREAD

**Issue by EderSantana** [fchollet/keras#1053]
@fchollet  just merged the code to reproduce some results of the Neural Turing Machines [paper](http://arxiv.org/pdf/1410.5401v2.pdf). This code also has some Keras lessons that the users might find interesting:

[Main code](https://github.com/fchollet/keras/blob/master/keras/layers/ntm.py)
[Companion example](https://github.com/fchollet/keras/blob/master/examples/neural_turing_machine_copy.py)

1) Sequence to Sequence learning. In the companion example, we solve the Copy problem where the input is first presented and should be reproduced in the output after an end-of-sequence vector is presented.

2) We use a single model to read the inputs and produce the outputs. See the figure bellow from Karpathy's [blog post](http://karpathy.github.io/2015/05/21/rnn-effectiveness/). While the input is presented, we don't care to what is the output. After we are done presenting the input that should be copied, we show the end-of-sequence vector and only zeros afterwards. While the zeros are being input we start caring to the output and comparing it to the target values. We use `sample_weight` to keep unimportant outputs from interfering with the cost function. `sample_weight` can be passed as a parameter to `model.train_on_batch` as shown [here] (https://github.com/fchollet/keras/blob/master/examples/neural_turing_machine_copy.py#L119)

![screen shot 2015-11-20 at 9 53 01 pm](https://cloud.githubusercontent.com/assets/3902480/11316316/381c1c74-8fd1-11e5-80c2-fbe04b3800d0.png)

3) Neural Turing Machines use a mechanism of attention. They are basically LSTMs that can store entire vectors, instead of single scalars, to their cells. We don't write or read all the stored vectors at once, we actually multiply most of these vectors by (almost) zero, but for one "attended" vector that is multiplied by (almost) one. When we average all these rescaled vectors the result is approximately the attended vector itself. These scaler sum to one, in other words they are a probability distributions calculated with Softmax over the memory locations. See the main code for more details.

Here is a navigation guide to the main code:
Start reading the `get_output` method, which is basically a `scan` loop. 
Afterwards, read the `step` function that the `scan` loop calls. Here is what the step function does (courtesy of my thesis proposal, based on my interpretations of Graves et. al. paper):

1. Using h_tm1, update the read vector (aka attention distribution) r_t = f_r(r_tm1, h_tm1)
2. Read from memory m_t = read(r_t, M_tm1)
3. Using the input and the read vector, update the RNN controller state h_t = RNN(x_t, h_tm1)
4. Using new state h_t, update the write vector (another probability distribution) w_t = f_w(w_tm1, h_t)
5. Write to memory M_t = write(M_tm1, h_t, w_t).

Whenever we call a method inside the `step` function for one of the operations above, you just read its code knowing what it is supposed to do.

A practical lesson I got from implementing this code was: Don't try to write the full model on the first sit. First write and test all the functions that you are supposed to call inside the scan/for loop in a separate notebook/script. Take your time and make your model as modular as possible. For that case, I first wrote each step to calculate the reading and writing operations and tested them. When I felt satisfied, I simply wrapped everything in the loop in the form you see in the main code. Since Keras already provides the equations to update the states of LSTMs and GRUs I reused the code I needed with proper credit.

**Comment by EderSantana** [comment]
I just realized that the issue list is not the appropriate place for such documentation. Sorry for the inconvenience.

**Comment by mrwns** [comment]
What about an entry in the wiki?

**Comment by EderSantana** [comment]
I don't know if Keras has one... I was thinking about blog post or something.

**Comment by fchollet** [comment]
The wiki sounds like a good choice indeed. We haven't been using it so far, but in this case it would be fitting.

Can you edit it? https://github.com/fchollet/keras/wiki

**Comment by EderSantana** [comment]
@fchollet sure! I started something here https://github.com/fchollet/keras/wiki 
Can you see what I wrote? Does it have to be PR'ed? 
Let me know what you think.

**Comment by fchollet** [comment]
Seems to work fine, thanks! Apparently no need to PR anything.

**Comment by placebokkk** [comment]
hi ,@EderSantana . Thanks for your tutorial . I run your NTM example but does not get the reasonable results.

Here is the output of mine:
   1/4000 [..............................] - ETA: 12796s - NTM: 0.6956 - LSTM: 0.6956
NTM  test acc: 50.0615
LSTM test acc: 50.046
 501/4000 [==>...........................] - ETA: 9532s - NTM: 0.6371 - LSTM: 0.5052 
NTM  test acc: 50.404
LSTM test acc: 52.7
1001/4000 [======>.......................] - ETA: 8220s - NTM: 0.5959 - LSTM: 0.3448
NTM  test acc: 50.73275
LSTM test acc: 51.52075
1501/4000 [==========>...................] - ETA: 6880s - NTM: 0.5654 - LSTM: 0.2593
NTM  test acc: 50.4695
LSTM test acc: 52.33425
2001/4000 [==============>...............] - ETA: 5517s - NTM: 0.5407 - LSTM: 0.2074
NTM  test acc: 50.5025
LSTM test acc: 52.852
2501/4000 [=================>............] - ETA: 4119s - NTM: 0.5191 - LSTM: 0.1737
NTM  test acc: 50.88325
LSTM test acc: 51.7495
3001/4000 [=====================>........] - ETA: 2738s - NTM: 0.4997 - LSTM: 0.1489
NTM  test acc: 51.15925
LSTM test acc: 52.7085
3501/4000 [=========================>....] - ETA: 1367s - NTM: 0.4814 - LSTM: 0.1306
NTM  test acc: 51.39875
LSTM test acc: 51.80225
4000/4000 [==============================] - 10955s - NTM: 0.4641 - LSTM: 0.1159

**Comment by EderSantana** [comment]
@placebokkk thanks for the feedback, I'll check that out.
 I'll make sure I'm setting the right random seed. 

Just one thing, don't throw the learned model out. Try to continue training it with more samples. 
In my own experiments, That model was bad (50% acc) for several epochs and suddenly jumped to 
a better results (we only visualize every 500 epochs). So maybe depending on the initialization,
it may take longer to converge (Graves et. al. reports using 1M updates before visualizing the model on the test set!).

Can you show the image of the last visualization? The one with the position of read and write heads?

**Comment by EderSantana** [comment]
@placebokkk I just tested it, but I seems to be just a matter of how long you train the method. Sometimes I get it to work as I said, sometimes it takes 10k grad updates. 

Also I there this one line of code that be affecting things or may not. I'll do a PR anyway. If you running your model in try increasing

**Comment by keunwoochoi** [comment]
The two links in the wiki are broken! @EderSantana

**Comment by billwiliams** [comment]
@EderSantana can we expect a keras 2.0.* compatible NTM ?
